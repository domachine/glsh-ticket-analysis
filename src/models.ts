export interface TextAnalysis {
  sentiment: number
  language: string | null
  keyPhrases: string[]
}

export interface Ticket {
  id: string
  title: string
  description: string
  priority: string
  userDefinedFields: {
    [key: string]: string
  }
}
