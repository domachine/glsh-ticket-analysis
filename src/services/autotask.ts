import { compose } from 'ramda'
import { Ticket } from '../models'
import { Params } from '../services'

type TicketFields = 'Title' | 'Description' | 'Priority'

export interface AutotaskUserDefinedField {
  Name: string[]
  Value?: string[]
}

export interface AutotaskUserDefinedFields {
  UserDefinedField: { Name: string[]; Value?: string[] }[]
}

export interface AutotaskTicket {
  Entity: {
    Title?: { _: string }[]
    Description?: { _: string }[]
    Priority?: { _: string }[]
    UserDefinedFields?: (string | AutotaskUserDefinedFields)[]
  }[]
}

export const getTicketField = (key: TicketFields) => (
  ticket: AutotaskTicket
): string | null => {
  const field = ticket.Entity[0][key]
  if (!field) return null
  return field[0]._
}

export const getTicketUserDefinedFields = (
  ticket: AutotaskTicket
): { [key: string]: string } => {
  const entity = ticket.Entity[0]
  if (!entity.UserDefinedFields) return {}
  const fieldWrapper = entity.UserDefinedFields[0] as AutotaskUserDefinedFields
  if (!fieldWrapper) return {}
  const fields = fieldWrapper.UserDefinedField
  if (!fields) return {}
  return fields.reduce((fields, field) => {
    if (field.Value === undefined) return fields
    return {
      ...fields,
      [field.Name[0]]: field.Value[0]
    }
  }, {})
}

export const updateTicketFields = (ticket: Ticket) =>
  compose(
    updateTicketField('Title', ticket.title),
    updateTicketField('Description', ticket.description),
    updateTicketField('Priority', ticket.priority),
    ...Object.keys(ticket.userDefinedFields).map(key =>
      updateTicketUserField(key, ticket.userDefinedFields[key])
    )
  )

export const updateTicketField = (name: TicketFields, value: string) => (
  ticket: AutotaskTicket
) => {
  const field = (ticket.Entity[0][name] || [])[0]
  return {
    ...ticket,
    Entity: [
      { ...ticket.Entity[0], [name]: [{ ...field, _: value }] },
      ...ticket.Entity.slice(1)
    ]
  }
}

export const updateTicketUserField = (name: string, value: string) => (
  ticket: AutotaskTicket
) => {
  const entity = ticket.Entity[0]
  const normalizedFields: AutotaskUserDefinedFields['UserDefinedField'] =
    !entity.UserDefinedFields || entity.UserDefinedFields[0] === ''
      ? []
      : (entity.UserDefinedFields[0] as AutotaskUserDefinedFields)
          .UserDefinedField
  const updatedFields = normalizedFields.find(field => field.Name[0] === name)
    ? normalizedFields.map(
        field => (field.Name[0] === name ? { ...field, Value: [value] } : field)
      )
    : [...normalizedFields, { Name: [name], Value: [value] }]
  return {
    ...ticket,
    Entity: [
      {
        ...ticket.Entity[0],
        UserDefinedFields: [{ UserDefinedField: updatedFields }]
      },
      ...ticket.Entity.slice(1)
    ]
  }
}

export async function fetchAutotaskTicket(
  this: Pick<Params, 'autotask'>,
  id: string
): Promise<AutotaskTicket | null> {
  const res = await this.autotask.query({
    queryxml: {
      entity: 'ticket',
      query: { field: { _: 'id', expression: { $: { op: 'equals' }, _: id } } }
    }
  })
  if (!res.EntityResults.length || !res.EntityResults[0].Entity) return null
  const autotaskTicket = res.EntityResults[0]
  return autotaskTicket
}

export async function fetchTicket(
  this: Pick<Params, 'autotask'>,
  id: string
): Promise<Ticket | null> {
  const autotaskTicket = await fetchAutotaskTicket.call(this, id)
  if (!autotaskTicket) return null
  return {
    id,
    title: getTicketField('Title')(autotaskTicket) || '',
    description: getTicketField('Description')(autotaskTicket) || '',
    priority: getTicketField('Priority')(autotaskTicket) || '-1',
    userDefinedFields: getTicketUserDefinedFields(autotaskTicket)
  }
}

export async function updateTicket(
  this: Pick<Params, 'autotask'>,
  ticket: Ticket
) {
  const autotaskTicket = await fetchAutotaskTicket.call(this, ticket.id)
  await this.autotask.update(updateTicketFields(ticket)(autotaskTicket))
}
