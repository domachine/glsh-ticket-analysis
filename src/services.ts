import { RequiredUriUrl, RequestAPI } from 'request'
import { RequestPromiseOptions, RequestPromise } from 'request-promise-native'
import { fetchTicket, updateTicket } from './services/autotask'
import { analyseText } from './services/azure-cognitive'
import { TextAnalysis, Ticket } from './models'

export interface Services {
  fetchTicket(id: string): Promise<Ticket>
  analyseText(text: string): Promise<TextAnalysis>
  updateTicket(ticket: Ticket): Promise<void>
}

export interface Params {
  autotask: any
  azureCognitive: RequestAPI<
    RequestPromise,
    RequestPromiseOptions,
    RequiredUriUrl
  >
}

export default (params: Params): Services => ({
  fetchTicket: fetchTicket.bind(params),
  updateTicket: updateTicket.bind(params),
  analyseText: analyseText.bind(params)
})
