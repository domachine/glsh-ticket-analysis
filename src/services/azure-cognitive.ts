import { Params } from '../services'

export class UnsupportedLanguageError extends Error {
  constructor(public language: string, public message: string) {
    super(message)
    this.name = UnsupportedLanguageError.name
  }
}

const SUPPORTED_LANGUAGES = new Set([
  'ar',
  'da',
  'de',
  'el',
  'en',
  'es',
  'fi',
  'fr',
  'it',
  'ja',
  'nl',
  'no',
  'pl',
  'pt-PT',
  'ru',
  'sv',
  'tr',
  'zh-Hans'
])

export const extractDocumentLanguage = (res: {
  detectedLanguages: { iso6391Name: string; score: number }[]
}): string | null => {
  const lang = res.detectedLanguages
    .slice()
    .sort((a, b) => a.score - b.score)[0]
  return lang ? lang.iso6391Name : null
}

export async function analyseText(this: Params, text: string) {
  const languagesRes = await this.azureCognitive.post(
    '/text/analytics/v2.0/languages',
    { json: { documents: [{ id: '1', text }] } }
  )
  if (languagesRes.errors[0]) throw new Error(languagesRes.errors[0].message)
  const language = extractDocumentLanguage(languagesRes.documents[0])
  if (!language) throw new Error("Language wasn't recognized")
  if (!SUPPORTED_LANGUAGES.has(language))
    throw new UnsupportedLanguageError(
      language,
      'Language is not supported by azure'
    )
  const [sentimentRes, keyPhrasesRes] = await Promise.all([
    this.azureCognitive.post('/text/analytics/v2.0/sentiment', {
      json: { documents: [{ id: '1', language, text }] }
    }),
    this.azureCognitive.post('/text/analytics/v2.0/keyPhrases', {
      json: { documents: [{ id: '1', language, text }] }
    })
  ])
  const errors = [sentimentRes, keyPhrasesRes]
    .map(res => res.errors[0])
    .filter(Boolean)
  if (errors[0]) throw new Error(errors[0].message)
  const sentiment = sentimentRes.documents[0].score
  const keyPhrases = keyPhrasesRes.documents[0].keyPhrases
  return { language, sentiment, keyPhrases }
}
