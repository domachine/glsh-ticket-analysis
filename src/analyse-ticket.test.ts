import {
  analyseTicket,
  calculatePriority,
  calculateTicketPriority
} from './analyse-ticket'
import { Ticket } from './models'
import { UnsupportedLanguageError } from './services/azure-cognitive'

test('Calculate priority', () => {
  expect(calculatePriority(0.1)).toEqual(4)
  expect(calculatePriority(0.2)).toEqual(4)
  expect(calculatePriority(0.21)).toEqual(3)
  expect(calculatePriority(0.29)).toEqual(3)
  expect(calculatePriority(0.5)).toEqual(0)
})

test('Calculate ticket-priority', () => {
  expect(calculateTicketPriority(3, 4)).toEqual(4)
  expect(calculateTicketPriority(5, 4)).toEqual(5)
  expect(calculateTicketPriority(2, 2)).toEqual(2)
})

test('Analyse ticket', async () => {
  const ticketId = '10708'
  const ticketTitle = ''
  const ticketDescription = 'Blubasdasdad'
  const ticketPriority = '1'
  const ticket: Ticket = {
    id: ticketId,
    title: ticketTitle,
    description: ticketDescription,
    priority: ticketPriority,
    userDefinedFields: {}
  }
  const sentiment = 0.146332
  const mockedAzureRes = { sentiment }
  const services = {
    fetchTicket: jest.fn(() => Promise.resolve(ticket)),
    analyseText: jest.fn(() => Promise.resolve(mockedAzureRes)),
    updateTicket: jest.fn(() => Promise.resolve())
  }
  await analyseTicket(services as any, ticketId)
  expect(services.fetchTicket.mock.calls[0]).toMatchSnapshot()
  expect(services.analyseText.mock.calls[0]).toMatchSnapshot()
  expect(services.updateTicket.mock.calls[0]).toMatchSnapshot()
})

test('Analyse ticket with long description', async () => {
  const ticketId = '10708'
  const ticketTitle = ''
  const ticketDescription = Array.from({ length: 6000 })
    .map(() => 'x')
    .join('')
  const ticketPriority = '1'
  const ticket: Ticket = {
    id: ticketId,
    title: ticketTitle,
    description: ticketDescription,
    priority: ticketPriority,
    userDefinedFields: {}
  }
  const sentiment = 0.146332
  const mockedAzureRes = { sentiment }
  const services = {
    fetchTicket: jest.fn(() => Promise.resolve(ticket)),
    analyseText: jest.fn(() => Promise.resolve(mockedAzureRes)),
    updateTicket: jest.fn(() => Promise.resolve())
  }
  await analyseTicket(services as any, ticketId)
  expect(ticketDescription.slice(0, 5000).length).toEqual(5000)
  expect(services.analyseText.mock.calls[0]).toEqual([
    (ticketTitle + '. ' + ticketDescription).slice(0, 5000)
  ])
})

test('Stop on unsupported language', async () => {
  const ticketId = '10708'
  const ticketTitle = ''
  const ticketDescription = 'asasd'
  const ticketPriority = '1'
  const ticket: Ticket = {
    id: ticketId,
    title: ticketTitle,
    description: ticketDescription,
    priority: ticketPriority,
    userDefinedFields: {}
  }
  const services = {
    fetchTicket: jest.fn(() => Promise.resolve(ticket)),
    analyseText: jest.fn(() =>
      Promise.reject(new UnsupportedLanguageError('ro', ''))
    ),
    updateTicket: jest.fn(() => Promise.resolve())
  }
  expect(await analyseTicket(services as any, ticketId)).toEqual({
    skip: true,
    message: ''
  })
  expect(services.updateTicket.mock.calls.length).toBe(0)
})
