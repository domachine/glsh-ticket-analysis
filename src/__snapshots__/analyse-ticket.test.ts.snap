// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Analyse ticket 1`] = `
Array [
  "10708",
]
`;

exports[`Analyse ticket 2`] = `
Array [
  ". Blubasdasdad",
]
`;

exports[`Analyse ticket 3`] = `
Array [
  Object {
    "description": "Blubasdasdad",
    "id": "10708",
    "priority": "4",
    "title": "",
    "userDefinedFields": Object {
      "keyphrases": "",
      "pbta": "1",
      "textanalytics": "15",
    },
  },
]
`;
